### PR TITLE
bpo-42653: Add constants to socket module for IsoTp CAN protocol

### PR DESCRIPTION
**Pull request by ghost**
This PR adds constants to socket module for IsoTp CAN protocol.

IsoTp CAN protocol is mainline since Kernel v5.10.
The basic socket constant for IsoTp has been available since Python 3.7 (August 2018),
so it could be used for default use-cases without having issues.

However several constants from isotp.h have not found their way into socket module yet. 
When using advanced options like padding, this flaw becomes visible.
https://gitlab.com/Menschel/socketcan/-/blob/master/socketcan/socketcan.py#L18

This PR may be related to https://github.com/python/cpython/pull/23794.

https://bugs.python.org/issue42653

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@menschel

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by menschel** [comment]
Dear Reviewer,
this PR apparently targets the same topic as https://github.com/python/cpython/pull/23794
Please advise.

**Comment by rumpelsepp** [comment]
My pullrequest seems to lack these `*_DEFAULT_*` constants. Are these new to the kernel module?

**Comment by menschel** [comment]
Yes,

https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/can/isotp.h

The one on @hartkopp 's repo is outdated.

**Comment by menschel** [comment]
Dear Reviewer the patch has been verified on a rpi3b.

Please also review my Unittest https://github.com/python/cpython/pull/26174/commits/1f7c2c54e8561930640fb1cfcd5c674e7805eb22 . It did not fail during make run before https://github.com/python/cpython/pull/26174/commits/2f19691b3558f6cee0a1254efef42af03e7f08a8 .

Thanks.

**Comment by menschel** [comment]
The patch also works on Python 3.9.5. This was build on a rpi0w.
```
$ python -i
Python 3.9.5 (heads/master-dirty:00ae35b, May 22 2021, 21:28:37) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> socket.CAN_ISOTP_OPTS
1
>>> socket.SOL_CAN_ISOTP
106
>>> 
```

File: Modules/socketmodule.c [hunk 1]
```diff
@@ -7809,7 +7809,10 @@ PyInit__socket(void)
     PyModule_AddIntMacro(m, CAN_ISOTP_FORCE_RXSTMIN);
     PyModule_AddIntMacro(m, CAN_ISOTP_RX_EXT_ADDR);
     PyModule_AddIntMacro(m, CAN_ISOTP_WAIT_TX_DONE);
+#ifdef CAN_ISOTP_SF_BROADCAST
+    /* This constant is new and not always available */
     PyModule_AddIntMacro(m, CAN_ISOTP_SF_BROADCAST);
```
Comment (rumpelsepp) [inline Modules/socketmodule.c:7814]: I assume the version of the kernel could be checked on buildtime, couldn't it?

Comment (ghost) [reply Modules/socketmodule.c:7814]: Technically yes, but there is no guarantee that this symbol really is there when the kernel version is above a specific version.
If you install the headers yourself with `linux-headers-$(uname -r)`, the assumption works.

But if the installation of headers has been obscured, like with Raspberry Pi 's package `raspberrypi-kernel-headers`, this assumption no longer works. It just adds another assumption that the packages 
`raspberrypi-kernel-headers` and `raspberrypi-kernel` are in sync.

Comment (rumpelsepp) [reply Modules/socketmodule.c:7814]: Personally, I would argue that the mainline upstream kernel is the source of truth and this must be fixed downstream. No idea how the CPython project deals with these kinds of issues.

Comment (ghost) [reply Modules/socketmodule.c:7814]: Well, lets wait for the Reviewer to comment on this. This should theoretically be  @tiran.

Comment (rumpelsepp) [reply Modules/socketmodule.c:7814]: aye.

**Comment by menschel** [comment]
BUMP.

This PR is stale for about a month now. Is there a way so accelerate this?

**Comment by MaxwellDupre** [comment]
Need to re-target Doc/whatsnew/3.10.rst to 3.12.

**Comment by hartkopp** [comment]
> Need to re-target Doc/whatsnew/3.10.rst to 3.12.

While you are at it:

Meanwhile _another_ new constant ` CAN_ISOTP_CF_BROADCAST` will hit the road in Linux 5.19 ...

https://github.com/pylessard/python-can-isotp/issues/67#issuecomment-1126349181

**Comment by menschel** [comment]
Be aware, these PRs have been stale for over a year.

I consider them dead.

**Comment by hartkopp** [comment]
> Be aware, these PRs have been stale for over a year. I consider them dead.

Yes but when @MaxwellDupre makes the zombie run again ... ;-)

**Pull request closed without merging**